// ng-zorro-antd/esm2015/space/space.module.js
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { BidiModule } from '@angular/cdk/bidi';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NzSpaceItemLegacyComponent } from './space-item.component';
import { NzSpaceItemDirective } from './space-item.directive';
import { NzSpaceComponent } from './space.component';
import * as ɵngcc0 from '@angular/core';
export class NzSpaceModule {
}
NzSpaceModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NzSpaceModule });
NzSpaceModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NzSpaceModule_Factory(t) { return new (t || NzSpaceModule)(); }, imports: [[BidiModule, CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NzSpaceModule, { declarations: function () { return [NzSpaceComponent, NzSpaceItemLegacyComponent, NzSpaceItemDirective]; }, imports: function () { return [BidiModule, CommonModule]; }, exports: function () { return [NzSpaceComponent, NzSpaceItemLegacyComponent, NzSpaceItemDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NzSpaceModule, [{
        type: NgModule,
        args: [{
                declarations: [NzSpaceComponent, NzSpaceItemLegacyComponent, NzSpaceItemDirective],
                exports: [NzSpaceComponent, NzSpaceItemLegacyComponent, NzSpaceItemDirective],
                imports: [BidiModule, CommonModule]
            }]
    }], null, null); })();
