// ng-zorro-antd/esm2015/avatar/avatar-group.component.js
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { ChangeDetectionStrategy, Component } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class NzAvatarGroupComponent {
}
NzAvatarGroupComponent.ɵfac = function NzAvatarGroupComponent_Factory(t) { return new (t || NzAvatarGroupComponent)(); };
NzAvatarGroupComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NzAvatarGroupComponent, selectors: [["nz-avatar-group"]], hostAttrs: [1, "ant-avatar-group"], exportAs: ["nzAvatarGroup"], ngContentSelectors: _c0, decls: 1, vars: 0, template: function NzAvatarGroupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NzAvatarGroupComponent, [{
        type: Component,
        args: [{
                selector: 'nz-avatar-group',
                exportAs: 'nzAvatarGroup',
                template: `
    <ng-content></ng-content>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush,
                host: {
                    class: 'ant-avatar-group'
                }
            }]
    }], null, null); })();
